CTPPS: raw-to-digi for diamond detectors

jan-kaspar opened the pull request:
Changes to include raw-to-digi code for the diamond detectors of CTPPS.

 * Built on top of #16010 .
 * Added `CTPPSDiamondDigi` class.
 * Added XML files with "cabling" mapping for diamonds.
 * "Cabling" mapping stored in EventSetup separately for each CTPPS subdetector.
 * Added `RawToDigiConverter::Run` for diamonds.
 * Added `ctppsRawToDigi` sequence to include all CTPPS modules with proper configuration.
 * Updated `EventContent` to include also the diamond products.
 * Added event counter to the `TotemVFATStatus` for DQM and debugging.

Below, a summary of tests performed.

llvm-analyzer: no warnings in CTPPS code.

CTPPS configs:
 * EventFilter/CTPPSRawToDigi/test/test_standard_sequence_cfg.py
 * RecoCTPPS/Configuration/test/raw_data_test.py
 * DQM/CTPPS/test/dqm_test_cfg.py

Si strip reconstruction (local tracks), one reference run
 * identical before and after: http://jkaspar.web.cern.ch/jkaspar/share/github/cmp.pdf

Global CMSSW configs:
 * DQM/Integration/python/clients/ctpps_dqm_sourceclient-live_cfg.py
 * Configuration/StandardSequences/python/RawToDigi_cff.py
   * tested to include the right CTPPS modules given the Run2_2016 era
 * Configuration/StandardSequences/python/Reconstruction_cff.py
   * tested to include the right CTPPS modules given the Run2_2016 era

Matrix test
 * `runTheMatrix.py -l limited` yielded:
   * 14 14 13 11 7 1 1 1 tests passed, 0 0 0 0 0 0 0 0 failed

jan-kaspar committed:
Removed obsolete T1/T2 code. Removed subSystem field from TotemSymbID
(subSystem information is now encoded in symbolicID).

jan-kaspar committed:
Removed trigger-VFAT code not used in CTPPS.

jan-kaspar committed:
Prepared for the diamond code.

setesami committed:
adding diamond mapping, diamond digi

jan-kaspar committed:
Printing moved to analyze method.

jan-kaspar committed:
More useful printout.

jan-kaspar committed:
Silently skip known but unsed node types.

jan-kaspar committed:
Frame position using full FED ID (old scheme supported as legacy only).

jan-kaspar committed:
Paths corrected.

jan-kaspar committed:
Comparison corrected.

jan-kaspar committed:
Safer and more flexible hierarchy control.

setesami committed:
diamond mapping implemented

setesami committed:
improvements for diamond digi

setesami committed:
remove extra comment and variable

setesami committed:
new improvements to diamond digi

setesami committed:
adding more methods to handle timing information

setesami committed:
improvment in comparison operator of digi class for diamond

setesami committed:
adding new class to handle diamond information in VFAT frame, raw to digi conversion implemented for diamond

setesami committed:
some cout to findout the ID problem of firmware

setesami committed:
new xml for mapping of diamond

setesami committed:
Inverting leading and trailing edges

setesami committed:
mismatching condition released for diamond, clean up for diamondvaftframe class

setesami committed:
revert the changes for id condition, relax this cut in py file

setesami committed:
add cout for mismatching problem of id

jan-kaspar committed:
Switched to the latest mapping by Nicola.

jan-kaspar committed:
Started restructuring the config files.

jan-kaspar committed:
Defined CTPPS raw-to-digi sequence.

jan-kaspar committed:
Test configs for strips only, diamonds only and standard sequence (both).

jan-kaspar committed:
DAQ mapping stored with sub-system label.

jan-kaspar committed:
Clean up.

jan-kaspar committed:
Separate mapping sources for strips and diamonds.

jan-kaspar committed:
Consistent sub-system names (as in CTPPSDetId).

jan-kaspar committed:
Test configs save reco file with outputCommands.

jan-kaspar committed:
Merge remote-tracking branch 'official-cmssw/CMSSW_8_1_X' into ctpps_diamond_raw_to_digi

jan-kaspar committed:
Run-dependent mapping configuration.

jan-kaspar committed:
Merge remote-tracking branch 'origin/ctpps_id_update' into ctpps_diamond_raw_to_digi

jan-kaspar committed:
Config files updated to recent changes.

setesami committed:
improvement in finding plane Id

jan-kaspar committed:
Fixed config to print the DAQ mapping.

nminafra committed:
Correct leading/trailing bitmap

nminafra committed:
Merge branch 'ctpps_diamond_raw_to_digi' of github.com:CTPPS/cmssw into ctps_diamond_raw_to_digi

nminafra committed:
Revert "Fixed config to print the DAQ mapping."

This reverts commit 72b4952e4697f8c6f74791b36a17faf685e0dfd4.

nminafra committed:
test commit

nminafra committed:
test commit

nminafra committed:
solved git problem

nminafra committed:
Merge branch 'ctpps_diamond_raw_to_digi' of github.com:CTPPS/cmssw into ctps_diamond_raw_to_digi

nminafra committed:
Renamed methods, easier to understand

jan-kaspar committed:
Fixing warnings issued by llvm-analyzer.

jan-kaspar committed:
Defined EventContent for diamonds.

jan-kaspar committed:
DQM test config includes data from diamonds, too.

nminafra committed:
cleanup

nminafra committed:
Merge branch 'ctpps_diamond_raw_to_digi' of github.com:CTPPS/cmssw into ctps_diamond_raw_to_digi

jan-kaspar committed:
DQM config updated to use the new raw-to-digi sequence.

nminafra committed:
git conflict, solved

nminafra committed:
reverted to correct version

jan-kaspar committed:
Merge branch 'ctpps_diamond_raw_to_digi' of github.com:CTPPS/cmssw into ctpps_diamond_raw_to_digi

jan-kaspar committed:
Merge remote-tracking branch 'official-cmssw/CMSSW_8_1_X' into ctpps_diamond_raw_to_digi

nminafra committed:
added event counter to TotemVFATStatus

jan-kaspar committed:
Merge branch 'ctpps_diamond_raw_to_digi' of github.com:CTPPS/cmssw into ctpps_diamond_raw_to_digi

jan-kaspar committed:
Merge remote-tracking branch 'official-cmssw/CMSSW_8_1_X' into ctpps_diamond_raw_to_digi

cmsbuild commented:
A new Pull Request was created by  @jan-kaspar  for CMSSW_8_1_X.

It involves the following packages:

CondFormats/CTPPSReadoutObjects
Configuration/StandardSequences
DQM/CTPPS
DQM/Integration
DataFormats/CTPPSDetId
DataFormats/CTPPSDigi
DataFormats/FEDRawData
EventFilter/CTPPSRawToDigi
Geometry/VeryForwardGeometryBuilder
RecoCTPPS/Configuration
RecoCTPPS/TotemRPLocal


The following packages do not have a category, yet:

DataFormats/CTPPSDigi
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories.py to assign category

@ghellwig, @civanch, @Dr15Jones, @cvuosalo, @emeschi, @ianna, @mdhildreth, @dmitrijus, @cmsbuild, @ggovi, @franzoni, @cerminar, @slava77, @mommsen, @mmusich, @vanbesien, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @apfeiffer1, @batinkov, @wmtan, @makortel, @GiacomoSguazzoni, @tocheng, @VinInn, @Martin-Grunewald, @rovere, @threus, @dgulhan this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

slava77 commented:
given almost ready state of #16010 I think we better get that merged first and then look at the raw2digi features without the DetId changes.
The alternative is to close #16010 and restart here.

jan-kaspar commented:
@slava77 As you like, both options are fine with me.

ianna commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 524820356f1887d784abbd91d6a827fb67f1f543
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16225/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16225/summary.html

dmitrijus commented:
+1

ianna commented:
+1

cvuosalo commented:
@jan-kaspar: Could you please give me diamond data and instructions for how to test this PR with it?
Thanks.

jan-kaspar commented:
@cvuosalo I suggest to run this config file:
```
  RecoCTPPS/Configuration/test/raw_data_test.py
```
It executes the standard raw-to-digi and local-reconstruction sequences on two raw-data files. The first has only strip data, the other has both strip and diamond. You will get some error messages from "RawDataUnpacker" - this is understood, some read-out chips went nuts. The config will create file AOD.root, using standard EventContent definition for AOD. Opening it in ROOT browser, you can check the presence of the diamond digi data: `CTPPSDiamondDigiedmDetSetVector_...`. There are two expected ID groups: around 2054E6 and 2071E6.

The two files cover both pre-TS2 and post-TS2 "eras" so the proposed config also checks that the "cabling" mapping EventSetup source switches the mappings correctly.

slava77 commented:
@davidlange6 
review of this PR is waiting for #16010 to get merged

jan-kaspar committed:
Merge remote-tracking branch 'official-cmssw/CMSSW_8_1_X' into ctpps_diamond_raw_to_digi

jan-kaspar commented:
Merge conflicts resolved.

cmsbuild commented:
Pull request #16471 was updated. @ghellwig, @civanch, @Dr15Jones, @cvuosalo, @emeschi, @ianna, @mdhildreth, @dmitrijus, @cmsbuild, @ggovi, @franzoni, @cerminar, @slava77, @mommsen, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

a bit more linear history would be nice (something for the future considerations)

cmsbuild commented:
The tests are being triggered in jenkins.

jan-kaspar commented:
@slava77 Yes, we will try to do better next time.

cmsbuild commented:
-1

Tested at: 7191f14d602db0e9d01298f9b9c29f0003dd1579


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7191f14d602db0e9d01298f9b9c29f0003dd1579
https://github.com/cms-sw/cmssw/commit/f8d0176944727706b8b7985f211399a1a65071c5
https://github.com/cms-sw/cmssw/commit/14cb42865bd11142ba7cbd35da36a925be73e11d
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16349/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16349/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16349/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
---> test TestFWCoreFrameworkGlobalStreamOne had ERRORS



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7191f14d602db0e9d01298f9b9c29f0003dd1579
https://github.com/cms-sw/cmssw/commit/f8d0176944727706b8b7985f211399a1a65071c5
https://github.com/cms-sw/cmssw/commit/14cb42865bd11142ba7cbd35da36a925be73e11d
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16349/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16349/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16349/summary.html

jan-kaspar commented:
I've checked the log for the two failing test units. For `testRecoMETMETProducers` the problem seems connected to this line
```
Failed to open the file 'ttbarForMetTests.root'
```
For `TestFWCoreFrameworkGlobalStreamOne` the reported problem seems to be
```
Failure using test_no_concurrent_module_cfg.py: status 1
```
Although I don't understand much, I feel that these problems are not related to this PR. Could you please check/confirm? Thanks!

slava77 commented:
On 11/16/16 12:51 AM, jan-kaspar wrote:
> I've checked the log for the two failing test units. For
> |testRecoMETMETProducers| the problem seems connected to this line
> 
> |Failed to open the file 'ttbarForMetTests.root' |
> 
> For |TestFWCoreFrameworkGlobalStreamOne| the reported problem seems to be
> 
> |Failure using test_no_concurrent_module_cfg.py: status 1 |
> 
> Although I don't understand much, I feel that these problems are not
> related to this PR. Could you please check/confirm? Thanks!


the unit test failures do not look related to this PR


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16471#issuecomment-260888447>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcboniXyllHMZHnXFgSro7RfTuOgjXks5q-sP8gaJpZM4KpkPQ>.
>

ianna commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 7191f14d602db0e9d01298f9b9c29f0003dd1579

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16463/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
23224.0 step2
<pre>runTheMatrix-results/23224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2023D5_GenSimHLBeamSpotFull+DigiFull_2023D5+RecoFullGlobal_2023D5+HARVESTFullGlobal_2023D5/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2023D5_GenSimHLBeamSpotFull+DigiFull_2023D5+RecoFullGlobal_2023D5+HARVESTFullGlobal_2023D5.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

slava77 commented:
@cmsbuild please test

the 2023 PR should work now

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7191f14d602db0e9d01298f9b9c29f0003dd1579
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16487/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16487/summary.html

ianna commented:
+1

dmitrijus commented:
+1

mommsen commented:
+1

cvuosalo commented on EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h:
```diff
@@ -0,0 +1,60 @@
+{
+  
+  public:
+    DiamondVFATFrame(const word* _data = NULL)
```
No name should have a leading underbar. Maybe change `_data` to `inputData`, but it's not essential.

setesami replied:
Done.

cvuosalo commented on EventFilter/CTPPSRawToDigi/python/ctppsRawToDigi_cff.py:
```diff
@@ -0,0 +1,66 @@
+import FWCore.ParameterSet.Config as cms
+
+# trigger data
+from EventFilter.CTPPSRawToDigi.totemTriggerRawToDigi_cfi import *
```
Is `import *` actually necessary? It would be better to explicitly name the modules you need to import.

cvuosalo commented:
I can see no real change in  EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc. Please remove it from this PR.

cvuosalo commented on RecoCTPPS/Configuration/test/raw_data_test.py:
```diff
@@ -4,39 +4,45 @@
 process.source = cms.Source("PoolSource",
-    fileNames = cms.untracked.vstring('file:/afs/cern.ch/user/j/jkaspar/public/run273062_ls0001-2_stream.root')
+  fileNames = cms.untracked.vstring(
+    'file:/afs/cern.ch/user/j/jkaspar/public/run273062_ls0001-2_stream.root',
```
Is the lifetime of the diamond detectors so short that they won't outlive this personal file under your home directory? Shouldn't we have a more standard file here?

slava77 commented on DataFormats/CTPPSDigi/interface/HPTDCErrorFlags.h:
```diff
@@ -0,0 +1,94 @@
+
+// namespace HPTDCErrors { 
+struct HPTDCErrorFlags {
+HPTDCErrorFlags(unsigned short flags=0): error_flags(flags) {;}
```
``` {} ``` instead of ```{;} ```

setesami replied:
Done.

slava77 commented on DataFormats/CTPPSDigi/interface/HPTDCErrorFlags.h:
```diff
@@ -0,0 +1,94 @@
+      case 11: return getHitLostROFifoOverflowGroup1();
+      case 12: return getHitErrorGroup0();
+      case 13: return getHitLostL1OverflowGroup0();
+      case 14: return getHitLostROFifoOverflowGroup0();
```
better use constants instead of magic numbers in 3 places

slava77 commented on DataFormats/CTPPSDigi/src/classes.h:
```diff
@@ -14,6 +14,7 @@
+#include "DataFormats/CTPPSDigi/interface/CTPPSDiamondDigi.h"
 
 #include <vector>
 
```
please change ``` namespace DataFormats_DataFormats ``` to ```namespace DataFormats_CTPPSDigi ```
looks like I missed this in the past iterations

jan-kaspar commented:
@setesami Could you please address the comments for the following files?
  * EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h
  * DataFormats/CTPPSDigi/interface/CTPPSDiamondDigi.h
  * DataFormats/CTPPSDigi/interface/HPTDCErrorFlags.h
  * DataFormats/CTPPSDigi/src/classes.h
  * DataFormats/CTPPSDigi/src/classes_def.xml

I will take care of the rest.

jan-kaspar committed:
Reverting unintended changes in RawDataUnpacker.

jan-kaspar committed:
Import * replaced with module names.

cmsbuild commented:
Pull request #16471 was updated. @ghellwig, @civanch, @Dr15Jones, @cvuosalo, @arunhep, @emeschi, @ianna, @mdhildreth, @dmitrijus, @cmsbuild, @ggovi, @franzoni, @cerminar, @slava77, @mommsen, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

slava77 commented:
... I suppose, we wait for edits from @setesami to continue with tests/review

jan-kaspar commented:
@cvousalo I've fixed
  * `ctppsRawToDigi_cff.py` has no more `import *`
  * reverted the changes in `RawDataUnpacker.cc`

However, I'm not sure what is a good replacement for the data file in my private directory. First, I tried paths like
```
/store/express/Run2016B/ExpressPhysics/FEVT/Express-v2/000/274/388/00000
```
but the directories seem empty for runs before TS1, so the life time isn't what is needed. Then, I tried looking for RAW data in DAS, yielding files like
```
/store/data/Run2016H/DoubleEG/RAW/v1/000/283/877/00000/00B1E52F-889B-E611-B4B9-FA163EA8BEB3.root
```
However, these data are typically on T0/T1 with "no user access". Could you please suggest a solution?

davidlange6 commented:
> On Nov 24, 2016, at 11:57 AM, jan-kaspar <notifications@github.com> wrote:
> 
> @cvousalo I've fixed
> 
> 	• ctppsRawToDigi_cff.py has no more import *
> 	• reverted the changes in RawDataUnpacker.cc
> However, I'm not sure what is a good replacement for the data file in my private directory. First, I tried paths like
> 
> /store/express/Run2016B/ExpressPhysics/FEVT/Express-v2/000/274/388/00000
> 
> but the directories seem empty for runs before TS1, so the life time isn't what is needed. Then, I tried looking for RAW data in DAS, yielding files like
> 
> /store/data/Run2016H/DoubleEG/RAW/v1/000/283/877/00000/00B1E52F-889B-E611-B4B9-FA163EA8BEB3.root
> 
> However, these data are typically on T0/T1 with "no user access". Could you please suggest a solution?
> 

You can access them with AAA. (and there is some crab fair-share on T1 and a lot on T2_CH_CERN)

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
@fabozzi @hengne 
We will need a relval matrix setup with 2016H, preferably one of the later runs with CTPPS diamonds included as well.
I'm not sure if any of the C/D/E can be dropped to keep the set smaller.
This will allow regular monitoring as well as a lock on a few files to be available on disk for testing.

@smuzaffar

mmusich commented:
@fabozzi @hengne @slava77 @davidlange6 we'll need to see how to get in this anticipated 2016H relval setup the correct conditions. 
Last update at https://github.com/cms-sw/cmssw/pull/16656 covers only up to run2016G (Sept2016 re-reco), and we still don't have the whole set of re-reco quality conditions for 2016H (to be delivered in the legacy re-reco). IMHO there are two options:
   * either we use a (snapshotted) version of the Prompt GT
   * or we live with the consequences of non updated IOVs (see [here](https://hypernews.cern.ch/HyperNews/CMS/get/calibrations/2748.html) )

Patching the current `run2_data` GT with prompt conditions for 2016H onwards is definitely too operationally intensive.  
@franzoni @arunhep

davidlange6 commented:
A snapshot from the prompt GT sounds good...

> On Nov 24, 2016, at 12:24 PM, Marco Musich <notifications@github.com> wrote:
> 
> @fabozzi @hengne @slava77 @davidlange6 we'll need to see how to get in this anticipated 2016H relval setup the correct conditions.
> Last update at #16656 covers only up to run2016G (Sept2016 re-reco), and we still don't have the whole set of re-reco quality conditions for 2016H (to be delivered in the legacy re-reco). IMHO there are two options:
> 
> 	• either we use a (snapshotted) version of the Prompt GT
> 	• or we live with the consequences of non updated IOVs (see here )
> Patching the current run2_data GT with prompt conditions for 2016H onwards is definitely too operationally intensive.
> @franzoni @arunhep
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jan-kaspar committed:
Using standard input data for testing.

cmsbuild commented:
Pull request #16471 was updated. @ghellwig, @civanch, @Dr15Jones, @cvuosalo, @arunhep, @emeschi, @ianna, @mdhildreth, @dmitrijus, @cmsbuild, @ggovi, @franzoni, @cerminar, @slava77, @mommsen, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

jan-kaspar commented:
For `raw_data_test.py` I've found convenient files on T2 at CERN.

setesami committed:
some corrections applied

cmsbuild commented:
Pull request #16471 was updated. @ghellwig, @civanch, @Dr15Jones, @cvuosalo, @arunhep, @emeschi, @ianna, @mdhildreth, @dmitrijus, @cmsbuild, @ggovi, @franzoni, @cerminar, @slava77, @mommsen, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

setesami commented on DataFormats/CTPPSDigi/interface/CTPPSDiamondDigi.h:
```diff
@@ -0,0 +1,128 @@
+ * March 2016
+ */
+
+#include <boost/cstdint.hpp>
```
Done thanks.

setesami commented on DataFormats/CTPPSDigi/src/classes_def.xml:
```diff
@@ -33,4 +34,21 @@
+  <class name="edm::Wrapper<edm::DetSet<CTPPSDiamondDigi> >"/>
+  <class name="edm::Wrapper<edm::DetSetVector<CTPPSDiamondDigi> >"/>
+
+  <class name="HPTDCErrorFlags"/>
```
Version added.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cvuosalo commented:
A test of the following file containing Diamond with 200 events against baseline CMSSW_8_1_X_2016-11-19-1100 shows no problems.
```
/store/express/Run2016H/ExpressPhysics/FEVT/Express-v2/000/283/877/00000/4EE44B0E-2499-E611-A155-02163E011938.root
```
The new collections created by this PR are added to RECO event content, and the old, obsolete Totem collections are removed:
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
    533.2 ->         0.0       -533     OLDO  -0.02     TotemVFATStatusedmDetSetVector_totemRPRawToDigi_RP_reRECO.
     45.2 ->         0.0        -45     OLDO  -0.00     TotemFEDInfos_totemRPRawToDigi_RP_reRECO.
     84.4 ->         0.0        -84     OLDO  -0.00     TotemRPDigiedmDetSetVector_totemRPRawToDigi_RP_reRECO.
      0.0 ->        36.0         36     NEWO   0.00     TotemFEDInfos_ctppsDiamondRawToDigi_TimingDiamond_reRECO.
      0.0 ->        90.8         91     NEWO   0.00     CTPPSDiamondDigiedmDetSetVector_ctppsDiamondRawToDigi_TimingDiamond_reRECO.
      0.0 ->       548.0        548     NEWO   0.02     TotemVFATStatusedmDetSetVector_totemRPRawToDigi_TrackingStrip_reRECO.
      0.0 ->        46.1         46     NEWO   0.00     TotemFEDInfos_totemRPRawToDigi_TrackingStrip_reRECO.
      0.0 ->        85.2         85     NEWO   0.00     TotemRPDigiedmDetSetVector_totemRPRawToDigi_TrackingStrip_reRECO.
      0.0 ->        72.8         73     NEWO   0.00     TotemVFATStatusedmDetSetVector_ctppsDiamondRawToDigi_TimingDiamond_reRECO.
-------------------------------------------------------------
  2356238 ->     2356475        237             0.0     ALL BRANCHES
```
Validation plots also show the removal of the obsolete Totem quantities.

CPU time and memory usage don't change significantly.
```
Baseline
Max VSIZ 3350.09 on evt 177 ; max RSS 2764.97 on evt 200
Time av 5.08095 s/evt   max 25.7501 s on evt 1
M1 Time av 4.97708 s/evt   max 18.2756 s on evt 15
M8 Time av 5.05072 s/evt   max 18.2756 s on evt 15
M16 Time av 4.94994 s/evt   max 13.0021 s on evt 90

PR 16471
Max VSIZ 3353.08 on evt 177 ; max RSS 2726.04 on evt 170
Time av 5.16564 s/evt   max 13.7636 s on evt 1
M1 Time av 5.12244 s/evt   max 13.2573 s on evt 90
M8 Time av 5.20954 s/evt   max 13.2573 s on evt 90
M16 Time av 5.18863 s/evt   max 13.2573 s on evt 90
```
Processing the diamond data only adds an insignificant amount of CPU time:
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.00%         0.00 ms/ev ->         0.17 ms/ev ctppsDiamondRawToDigi
  ---------- ------------     --------                  ----       ------------
Job total:  4.76559 s/ev ==> 4.9357 s/ev
```
The test config for performing the Diamond data raw-to-digi runs successfully.

cmsbuild commented:
+1
Tested at: c0c2a48885f184a4ccd9c35066020e3315a22da9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16597/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16471/16597/summary.html

cvuosalo commented:
+1

For #16471 c0c2a48885f184a4ccd9c35066020e3315a22da9

For CT-PPS, adding raw-to-digi for Diamond data.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_1_X_2016-11-24-1100 show no problems, only the removal of obsolete Totem quantities in the Validation plots. A further test showed that the correct handling of the Diamond data with no significant negative impacts on CPU time, memory usage, or event content size (see https://github.com/cms-sw/cmssw/pull/16471#issuecomment-262851554)

slava77 commented:
90 bytes/event in CTPPSDiamondDigiedmDetSetVector_ctppsDiamondRawToDigi_TimingDiamond
is it a normal occupancy?
How many objects do we have there or is this 90bytes just empty collections?
I'm just trying to understand that we tested files with actual diamonds data.

jan-kaspar commented:
One should expect 56 DIGI records per event (2 RP * 4 planes * 7 channels). I've run 1000 events from the file Carl used:
```
/store/express/Run2016H/ExpressPhysics/FEVT/Express-v2/000/283/877/00000/4EE44B0E-2499-E611-A155-02163E011938.root
```
and get:
![diamond_id_histogram](https://cloud.githubusercontent.com/assets/17830858/20628539/e12cb8fe-b326-11e6-9281-013072d8b30f.png)
so diamond data are in.

I'm more surprised by the 90 B per event. Since the sizeof(DIGI class) is 16 B, I'd expect the collection to be more ~900 B. Could there be a kind of compression factor?

slava77 commented:
On 11/25/16 6:52 AM, jan-kaspar wrote:
> One should expect 56 DIGI records per event (2 RP * 4 planes * 7
> channels). I've run 1000 events from the file Carl used:
> 
> |/store/express/Run2016H/ExpressPhysics/FEVT/Express-v2/000/283/877/00000/4EE44B0E-2499-E611-A155-02163E011938.root
> |
> 
> and get:
> diamond_id_histogram
> <https://cloud.githubusercontent.com/assets/17830858/20628539/e12cb8fe-b326-11e6-9281-013072d8b30f.png>
> so diamond data are in.
> 
> I'm more surprised by the 90 B per event. Since the sizeof(DIGI class)
> is 16 B, I'd expect the collection to be more ~900 B. Could there be a
> kind of compression factor?

Carl and Jan
please post the uncompressed size in your tests. (edmEventSize -v
file.root | grep for the right branch)
Thanks



> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16471#issuecomment-262971565>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcblIydEw63die7fUhCoRoXZ4yjUbBks5rBvYrgaJpZM4KpkPQ>.
>

slava77 commented:
On 11/25/16 6:52 AM, jan-kaspar wrote:
> One should expect 56 DIGI records per event (2 RP * 4 planes * 7
> channels). I've run 1000 events from the file Carl used:
> 
> |/store/express/Run2016H/ExpressPhysics/FEVT/Express-v2/000/283/877/00000/4EE44B0E-2499-E611-A155-02163E011938.root
> |
> 
> and get:
> diamond_id_histogram
> <https://cloud.githubusercontent.com/assets/17830858/20628539/e12cb8fe-b326-11e6-9281-013072d8b30f.png>
> so diamond data are in.
> 
> I'm more surprised by the 90 B per event. Since the sizeof(DIGI class)
> is 16 B, I'd expect the collection to be more ~900 B. Could there be a
> kind of compression factor?

Actually, if the set of IDs is the same and the digi values are almost
all zero or something equivalently trivial,
then the compression factor should be high.
Was there any non-trivial data in the diamonds channels in this run?


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16471#issuecomment-262971565>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcblIydEw63die7fUhCoRoXZ4yjUbBks5rBvYrgaJpZM4KpkPQ>.
>

jan-kaspar commented:
```
File AOD.root Events 1000
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
CTPPSDiamondDigiedmDetSetVector_ctppsDiamondRawToDigi_TimingDiamond_CTPPSReconstructionChainTest. 1908.71 80.279
```

There are still the same 56 IDs and as far as I understand (I'm not expert on these detectors), there is no zero-suppression neither in DAQ neither in raw-to-digi. This probably justifies the large compression factor.

The amount of non-trivial data can be inferred from the rate of non-zero values of "leading edge time":
![leading_edge_time_histogram](https://cloud.githubusercontent.com/assets/17830858/20629389/9393902c-b32b-11e6-8638-f550e460dcd7.png)

slava77 commented:
On 11/25/16 7:27 AM, jan-kaspar wrote:
> |File AOD.root Events 1000 Branch Name | Average Uncompressed Size
> (Bytes/Event) | Average Compressed Size (Bytes/Event)
> CTPPSDiamondDigiedmDetSetVector_ctppsDiamondRawToDigi_TimingDiamond_CTPPSReconstructionChainTest.
> 1908.71 80.279 |
> 
> There are still the same 56 IDs and as far as I understand (I'm not
> expert on these detectors), there is no zero-suppression neither in DAQ
> neither in raw-to-digi. This probably justifies the large compression
> factor.
> 
> The amount of non-trivial data can be inferred from the rate of non-zero
> values of "leading edge time":
> leading_edge_time_histogram
> <https://cloud.githubusercontent.com/assets/17830858/20629389/9393902c-b32b-11e6-8638-f550e460dcd7.png>

Thanks for the plot of the times.
The sizes then look reasonable.


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16471#issuecomment-262979194>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcblieSpoz0lSWDA3rV62xryRiS0iDks5rBv5tgaJpZM4KpkPQ>.
>

cvuosalo commented:
@slava77: Here are the sizes of the Diamond collections from my test (https://github.com/cms-sw/cmssw/pull/16471#issuecomment-262851554):

```
File step3.root Events 200
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
CTPPSDiamondDigiedmDetSetVector_ctppsDiamondRawToDigi_TimingDiamond_reRECO. 1910.7 90.79
TotemVFATStatusedmDetSetVector_ctppsDiamondRawToDigi_TimingDiamond_reRECO. 1755.77 72.805
TotemFEDInfos_ctppsDiamondRawToDigi_TimingDiamond_reRECO. 77.955 36.01
```

ianna commented:
+1

jan-kaspar commented:
Is there something else we can do for this PR?

mommsen commented:
+1

dmitrijus commented:
+1

forthommel commented:
Dear AlCa/db/operations experts: Is there something else one can do for this PR?

mmusich commented:
+1

ggovi commented:
+1

slava77 commented:
@fabozzi 
this is in follow up to the ORP discussion
283877 data should should work for the tests

jan-kaspar commented:
Is there a reason why this PR is still pending? Thanks!

slava77 commented:
this PR is 81X, which has a rather narrow scope to be used in preparatory 2017 MC production.

at the time this PR was open, 81X was still the development release, but then it closed
and raw2digi is probably needed just in 80X and 90X now.

Please clarify your plans for 81X if they clearly exist.

jan-kaspar commented:
This PR was intended for the development release. Now I see that this PR has been cloned into #16616 and is already merged in CMSSW_9_0_X. I was not aware of that, sorry. So I will close this PR as it is not useful and make a back-port to 8_0_X instead.
